### PR TITLE
[DAI-149] : Review and Standardize AIPlatform API to v3

### PR DESCRIPTION
**Pull request by anidjerdor**

### CODE REVIEW

File: entities/bizaiChatsMessage.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      "example": "d290f1ee-6c54-4b01-90e6-d701748f0851"
+    },
+    "content": {
+      "type": "string",
+      "description": "content of the message",
+      "example": "Hello, how can I help you?"
+    },
+    "author_role": {
+      "type": "string",
+      "description": "role of the author of the message",
```
Comment (ramsterita) [inline entities/bizaiChatsMessage.json:16]: Make sure all descriptions start with Capital letter so we are consistent

File: entities/bizaiChats.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
+{
+  "type": "object",
+  "properties": {
+    "uid": {
+      "type": "string",
+      "description": "uid of the chat",
```
Comment (ramsterita) [inline entities/bizaiChats.json:6]: change uid to "The unique identifier..." in all places

File: entities/bizaiChats.json [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      "example": "d290f1ee-6c54-4b01-90e6-d701748f0851"
+    },
+    "actor_uid": {
+      "type": "string",
+      "description": "uid of the actor that created the chat",
+      "example": "ghp0f1ee-6c54-4b01-90e6-d701748f0851"
+    },
+    "agent": {
```
Comment (ramsterita) [inline entities/bizaiChats.json:14]: is'nt this a reference to a strict list of agents we support?

File: entities/bizaiChats.json [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      "type": "string",
+      "description": "Agent associated with the chat",
+      "example": "vanilla"
+    },
+    "metadata": {
```
Comment (ramsterita) [inline entities/bizaiChats.json:19]: Is this the system message? if it is than metadata is not a good name. lets discuss this

File: entities/bizaiChatsMessage.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (ramsterita) [inline entities/bizaiChatsMessage.json:1]: This should be bizaiChatMessage (not Chats...)

File: entities/bizaiChats.json [hunk 6]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (ramsterita) [inline entities/bizaiChats.json:1]: should be bizaiChat (not plural) only the end point is plural

File: swagger/Platfrom API/bizaiChats.json [hunk 7]
```diff
@@ -0,0 +1,407 @@
+{
+  "openapi": "3.0.0",
+  "info": {
+    "title": "Platform Management",
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:4]: change title to BizAI

File: swagger/Platfrom API/bizaiChats.json [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+            }
+          }
+        },
+        "security": [
+          {
+            "api_key": []
+          }
+        ],
+        "tags": [
+          "chatmng"
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:111]: The tag should be Chat not chatmg

File: swagger/Platfrom API/bizaiChats.json [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+        ],
+        "x-auth-type": "Directory"
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:113]: remove this line please

File: swagger/Platfrom API/bizaiChats.json [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+      },
+      "post": {
+        "operationId": "ChatsController_createChat",
+        "summary": "Create a chat",
+        "description": "Create a new chat session between a user and an agent",
+        "requestBody": {
+          "content": {
+            "application/json": {
+              "schema": {
+                "$ref": "https://vcita.github.io/developers-hub/entities/bizaiChats.json"
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:123]: after fixing the name, fix the reference

File: swagger/Platfrom API/bizaiChats.json [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+            "in": "path",
+            "required": true,
+            "schema": {
+              "type": "string"
+            },
+            "description": "The unique identifier of the chat"
+          }
+        ],
+        "responses": {
+          "200": {
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:177]: We need to add error responses too

File: swagger/Platfrom API/bizaiChats.json [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+          "chatmng"
+        ],
+        "x-auth-type": "Directory"
+      }
+    },
+    "/v3/bizai-chats/{uid}": {
+      "get": {
+        "operationId": "ChatsController_getChat",
+        "summary": "Retrieve a chat",
+        "description": "Get a specific chat session based on its unique identifier (UID)",
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:164]: Please add which tokens can access this endpoint (Business/Directory/Application)

File: swagger/Platfrom API/bizaiChats.json [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+        },
+        "security": [
+          {
+            "api_key": []
+          }
+        ],
+        "tags": [
+          "chatmng"
+        ],
+        "x-auth-type": "Directory"
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:204]: remove

File: swagger/Platfrom API/bizaiChats.json [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+      }
+    },
+    "/v3/bizai-chats/{uid}/messages": {
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:207]: do not nest the endpoint, it should be /v3/messages, you need to add the chat_uid to the path parameters

File: swagger/Platfrom API/bizaiChats.json [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+      "get": {
+        "operationId": "ChatsController_getMessages",
+        "summary": "List all messages",
+        "description": "Get all messages exchanged within a specific chat",
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:211]: add who can access this end point (see above)

File: swagger/Platfrom API/bizaiChats.json [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+          {
+            "in": "query",
+            "name": "sort",
+            "schema": {
+              "type": "string",
+              "default": "updated_at:desc"
+            },
+            "description": "Sort by fields, e.g., \"created_at:asc,updated_at:desc\"."
+          },
+          {
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:250]: We are stopping the support for LHS filters, the filters should be explicitly defined.

File: swagger/Platfrom API/bizaiChats.json [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+            "in": "query",
+            "name": "filter",
+            "schema": {
+              "type": "string"
+            },
+            "description": "Filter by a field. Example: filter[created_at][gt]=2019-01-01, filter[name]=my-name"
+          }
+        ],
+        "responses": {
+          "200": {
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:260]: missing error codes

File: swagger/Platfrom API/bizaiChats.json [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+        },
+        "security": [
+          {
+            "api_key": []
+          }
+        ],
+        "tags": [
+          "chatmng"
+        ],
+        "x-auth-type": "Directory"
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:307]: remove

File: swagger/Platfrom API/bizaiChats.json [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+              }
+            }
+          }
+        },
+        "security": [
+          {
+            "api_key": []
+          }
+        ],
+        "tags": [
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:304]: Tags are wrong they need to reflect a group of endpoints dealing with the same entity so tag for this should be BizAIMessage, and for the chat endpoints it should be BizAIChat
The reason behind it, is the way readme and swagger groups the endpoint in the documentation.
Please apply to all tags

File: swagger/Platfrom API/bizaiChats.json [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+            }
+          },
+          "required": true
+        },
+        "responses": {
+          "201": {
+            "description": "",
+            "content": {
+              "application/json": {
+                "schema": {
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:353]: This should be a reference, not a hard coded schema

**Review by ramsterita — CHANGES_REQUESTED** [review]

File: swagger/Platfrom API/bizaiChats.json [hunk 21]
```diff
@@ -0,0 +1,407 @@
+{
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:1]: We need to create a folder named BizAI and put this file in it

File: entities/bizaiChats.json [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+      "example": {
+        "directory_uid": "k98axtpqg7h1whh6",
+        "instruction": "You are a helpful assistant. Answer all questions to the best of your ability."
+      }
+    },
+    "created_at": {
+      "type": "string",
+      "description": "Date the chat was created",
+      "format": "date-time",
+      "example": "2024-05-20T14:00:00.000Z"
```
Comment (ramsterita) [inline entities/bizaiChats.json:31]: You don't need to have an example here and also at the end of the file. you can keep only the full example at the end

File: swagger/Platfrom API/bizaiChats.json [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+    "version": "1.0.0"
+  },
+  "servers": [
+    {
+      "url": "https://api.vcita.biz/v3/"
+    }
+  ],
+  "tags": [
+    {
+      "name": "chatmng",
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:16]: You need to have 2 tags BizAIChat and BizAIChatMessage and add them correspondingly to the appropriate endpoints

File: swagger/Platfrom API/bizaiChats.json [hunk 24]
```diff
@@ -0,0 +1,407 @@
...
+            "in": "query",
+            "name": "sort",
+            "schema": {
+              "type": "string",
+              "default": "updated_at:desc"
+            },
+            "description": "Sort by fields, e.g., \"created_at:asc,updated_at:desc\"."
+          },
+          {
+            "in": "query",
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:56]: You need to be explicit about supported filters and set them up as parameters. see documentation about filter (its different than the current standard.

File: swagger/Platfrom API/bizaiChats.json [hunk 25]
```diff
@@ -0,0 +1,407 @@
...
+            "name": "filter",
+            "schema": {
+              "type": "string"
+            },
+            "description": "Filter by a field. Example: filter[created_at][gt]=2019-01-01, filter[name]=my-name"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "A list of chats.",
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:66]: remove period at the end

File: swagger/Platfrom API/bizaiChats.json [hunk 26]
```diff
@@ -0,0 +1,407 @@
...
+        },
+        "security": [
+          {
+            "api_key": []
+          }
+        ],
+        "tags": [
+          "chatmng"
+        ],
+        "x-auth-type": "Directory"
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:157]: please remove the x-auth-type

File: swagger/Platfrom API/bizaiChats.json [hunk 27]
```diff
@@ -0,0 +1,407 @@
...
+        "description": "Get a list of all ongoing chat sessions",
+        "parameters": [
+          {
+            "in": "query",
+            "name": "page",
+            "schema": {
+              "type": "integer",
+              "default": 1
+            },
+            "description": "Page number of the results."
```
Comment (ramsterita) [inline swagger/Platfrom API/bizaiChats.json:34]: Remove period from all descriptions that are one sentence

**Review by ramsterita — CHANGES_REQUESTED** [review]

File: swagger/Platfrom API/bizaiChats.json (hunk 12 above)
Comment (anidjerdor) [reply swagger/Platfrom API/bizaiChats.json:164]: Can you provide an example? I didn't find it in the examples

File: swagger/Platfrom API/bizaiChats.json (hunk 11 above)
Comment (anidjerdor) [reply swagger/Platfrom API/bizaiChats.json:177]: This is missing in the example json, can you add it to the example?

File: swagger/Platfrom API/bizaiChats.json (hunk 16 above)
Comment (anidjerdor) [reply swagger/Platfrom API/bizaiChats.json:250]: Do you want me to remove the filter query?

Comment (ramsterita) [reply swagger/Platfrom API/bizaiChats.json:250]: yes, we only support explicit filters. things you defined and documented. There is a paragraph on this in the documentation

File: entities/bizaiChats.json (hunk 22 above)
Comment (ramsterita) [reply entities/bizaiChats.json:31]: You need to have one example object at the end. This is just redundant. I will not fail the PR for this, but it just creates clutter

File: swagger/Platfrom API/bizaiChats.json (hunk 23 above)
Comment (ramsterita) [reply swagger/Platfrom API/bizaiChats.json:16]: The tags will become the name of the API group that are tagged together. any endpoint with the the same tag will be grouped and named according to its tag. 
image
in this example you see the developer used "card request" and "cards as tags. We want to group messages CRUD + List and to group Chat CRUD +List

File: swagger/Platfrom API/bizaiChats.json (hunk 24 above)
Comment (ramsterita) [reply swagger/Platfrom API/bizaiChats.json:56]: Read this section in the documentation:
https://myvcita.atlassian.net/wiki/spaces/IT/pages/2941976643/API+v3+Standards+-+WIP#Filtering
I am not sure which example to you mean

Comment (anidjerdor) [reply swagger/Platfrom API/bizaiChats.json:56]: The `subscription.json` file

File: entities/BizAI/AIChatMessages.json
Comment (ramsterita) [inline entities/BizAI/AIChatMessages.json:1]: please rename this to singular AIChatMessage

File: entities/BizAI/AIChatStreamMessages.json
Comment (ramsterita) [inline entities/BizAI/AIChatStreamMessages.json:1]: please rename to singular

File: entities/BizAI/AIChats.json
Comment (ramsterita) [inline entities/BizAI/AIChats.json:1]: please rename to singular

File: entities/BizAI/AIChats.json [hunk 28]
```diff
@@ -0,0 +1,63 @@
+{
+  "type": "object",
+  "properties": {
+    "success": {
+      "type": "boolean"
+    },
```
Comment (ramsterita) [inline entities/BizAI/AIChats.json:6]: we dont need the response object in the entity. please remove it

**Review by ramsterita — COMMENTED** [review]
I added a few minor changes

File: entities/BizAI/AIChats.json (hunk 28 above)
Comment (anidjerdor) [reply entities/BizAI/AIChats.json]: GPT 😅

**Review by ramsterita — APPROVED** [review]

**Review by ramsterita — APPROVED** [review]

**Pull request merged**